vendor/github.com/containers/libhvee/pkg/ignition/kvp.go

```diff
@@ -0,0 +1,125 @@
+//go:build linux
+
+package ignition
+
+import (
+	"fmt"
+	"unsafe"
+
+	"golang.org/x/sys/unix"
+)
+
+// readKvpData reads all key-value pairs from the hyperv kernel device and creates
+// a map representation of them
+func readKvpData() (map[string]string, error) {
+	ret := make(map[string]string)
+
+	kvp, err := unix.Open(KvpKernelDevice, unix.O_RDWR|unix.O_CLOEXEC, 0)
+	if err != nil {
+		return nil, err
+	}
+	defer unix.Close(kvp)
+
+	var (
+		hvMsg    hvKvpMsg
+		hvMsgRet hvKvpMsgRet
+	)
+
+	const sizeOf = int(unsafe.Sizeof(hvMsg))
+
+	var (
+		asByteSlice    []byte = (*(*[sizeOf]byte)(unsafe.Pointer(&hvMsg)))[:]
+		retAsByteSlice []byte = (*(*[sizeOf]byte)(unsafe.Pointer(&hvMsgRet)))[:]
+	)
+
+	hvMsg.kvpHdr.operation = KvpOpRegister1
+
+	l, err := unix.Write(kvp, asByteSlice)
+	if err != nil {
+		return nil, err
+	}
+	if l != int(sizeOf) {
+		return nil, ErrUnableToWriteToKVP
+	}
+
+next:
+	for {
+		var pfd unix.PollFd
+		pfd.Fd = int32(kvp)
+		pfd.Events = unix.POLLIN
+		pfd.Revents = 0
+
+		howMany, err := unix.Poll([]unix.PollFd{pfd}, Timeout)
```

Comment:
`poll(2)` can have spurious wakeups, in which case we could presumably block on the following `read(2)`, since we haven't set the FD nonblocking.

Reply:
even with the timeout ?